### ISSUE TITLE
/etc/hosts appears to be assigned at random now

### ISSUE THREAD

**Issue by Techsture** [awslabs/deeplearning-cfn#13]
Whenever we spin up an environment with this template, the script `/opt/deeplearning/dl_cfn_setup_v2.py` gets called on the instances. This script updates the `/etc/hosts` file on each node with the list of all the addresses of the hosts in the cluster so it looks like this:

```
# cat /etc/hosts
127.0.0.1   localhost localhost.localdomain localhost4 localhost4.localdomain4
::1         localhost6 localhost6.localdomain6
172.31.4.18 deeplearning-master
172.31.4.18 deeplearning-worker1
172.31.17.209 deeplearning-worker2
172.31.23.139 deeplearning-worker3
172.31.28.203 deeplearning-worker4
172.31.39.46 deeplearning-worker5
172.31.0.67 deeplearning-worker6
172.31.4.223 deeplearning-worker7
172.31.40.227 deeplearning-worker8
```

Note that the address for `deeplearning-master` and `deeplearning-worker1` are the same. This is correct. The last few times I created a new environment, though, it's been assigning the hostnames based on their IP order. This means that the `deeplearning-master` could end up being assigned as any one of the other worker nodes.

The scripts being used internally by our engineers testing the distributed training systems on AWS rely on `deeplearning-master` being assigned as `deeplearning-worker1`.  Something changed recently that's stopping this from happening.

Has anyone else encountered this?  As far as I can tell, it's worked fine up until a few days ago.

**Comment by lupesko** [comment]
hmm. I can't see any recent changes that may have created this.
@nswamy can you take a look?
Marking as a bug for now.

**Comment by Techsture** [comment]
We've gotten lucky with the last environment we spun up in that the IP addresses assigned just happened to have the master numerically first as well.  I was considering putting in a provision that always put the master in the x.x.1.x subnet and the rest can go in x.x.2+.x, but haven't acted on that yet.

Thanks for taking a look!

**Comment by nswamy** [comment]
Did you modify the template, I am not sure how you got 172 addresses on your instances, the template actually sets up instances in subnets with CIDR block 10.0.x.x, see the below in the template
https://github.com/awslabs/deeplearning-cfn/blob/master/cfn-template/deeplearning.template#L126 
The master will always be in the subnet 10.0.0.0 and workers starting from 10.0.1.0, so sorting all the ips should still have the master as worker1.

If you have modified the template to suit your own use case and are having this issue, may be you can remove this line  
https://github.com/awslabs/deeplearning-cfn/blob/master/cfn-bootstrap/dl_cfn_setup_v2.py#L331
and add the line after 
https://github.com/awslabs/deeplearning-cfn/blob/master/cfn-bootstrap/dl_cfn_setup_v2.py#L342
`worker_instance_ips.insert(0, master_instance_ip)`

Let me know if that helps.

**Comment by Techsture** [comment]
Hi @nswamy,

Yes, I did have to modify it.  We had to use an existing EFS, which means we couldn't have this CFN continue to make a new VPC/EFS every time we ran it (copying 3+TB from S3 each time we spun up a new environment was taking awhile).

Our subnet config mapping looks like this:
`"SubnetConfig" : {
      "VPC"     : { "CIDR" : "172.31.0.0/16" },
      "Public"  : { "CIDR" : "172.31.0.0/24" },
      "Private" : { "CIDR" : "172.31.1.0/24" }
},`

Your comment helped me track down the problem, though.  I had changed the `MasterAutoScalingGroup` and `WorkerAutoScalingGroup` as follows:
`"VPCZoneIdentifier" : [ 
          "subnet-6cd24825",
          "subnet-dfb6f6b8",
          "subnet-8307f3d8"
],`

I don't know why I changed it to be that, but I'm sure that's why it's launching incorrectly now.

Thanks a lot for your help, I really appreciate it!
